common/task/weightedRoundRobinTaskScheduler.go

```diff
@@ -0,0 +1,195 @@
+	scheduler := &weightedRoundRobinTaskSchedulerImpl{
+		status:        common.DaemonStatusInitialized,
+		numPriorities: numPriorities,
+		taskChs:       make([]chan PriorityTask, numPriorities),
+		shutdownCh:    make(chan struct{}),
+		notifyCh:      make(chan struct{}, 1),
+		logger:        logger,
+		metricsScope:  metricsScope,
+		options:       options,
+		processor: NewParallelTaskProcessor(
+			logger,
+			metricsScope,
+			&ParallelTaskProcessorOptions{
+				QueueSize:   wRRTaskProcessorQueueSize,
+				WorkerCount: options.WorkerCount,
+				RetryPolicy: options.RetryPolicy,
+			},
+		),
+	}
+
+	for i := 0; i < numPriorities; i++ {
+		scheduler.taskChs[i] = make(chan PriorityTask, options.QueueSize)
+	}
+
+	return scheduler, nil
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&w.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	w.processor.Start()
+
+	w.dispatcherWG.Add(1)
+	go w.dispatcher()
+
+	w.logger.Info("Weighted round robin task scheduler started.")
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&w.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(w.shutdownCh)
+
+	w.processor.Stop()
+
+	if success := common.AwaitWaitGroup(&w.dispatcherWG, time.Minute); !success {
+		w.logger.Warn("Weighted round robin task scheduler timedout on shutdown.")
+	}
+
+	w.logger.Info("Weighted round robin task scheduler shutdown.")
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) Submit(task PriorityTask) error {
+	w.metricsScope.IncCounter(metrics.PriorityTaskSubmitRequest)
+	sw := w.metricsScope.StartTimer(metrics.PriorityTaskSubmitLatency)
+	defer sw.Stop()
+
+	priority := task.Priority()
+	if priority >= w.numPriorities {
+		return errors.New("task priority exceeds limit")
+	}
+	select {
+	case w.taskChs[priority] <- task:
+		select {
+		case w.notifyCh <- struct{}{}:
+			// sent a notification to the dispatcher
+		default:
+			// do not block if there's already a notification
+		}
+		return nil
+	case <-w.shutdownCh:
+		return ErrTaskSchedulerClosed
+	}
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) dispatcher() {
+	defer w.dispatcherWG.Done()
+
+	outstandingTasks := false
+
+	for {
+		if !outstandingTasks {
+			// if no task is dispatched in the last round,
+			// wait for a notification
+			select {
+			case <-w.notifyCh:
+				// block until there's a new task
+			case <-w.shutdownCh:
+				return
+			}
+		}
+
+		outstandingTasks = false
+		for priority := 0; priority != w.numPriorities; priority++ {
+			for i := 0; i != w.options.Weights[priority]; i++ {
+				select {
+				case task := <-w.taskChs[priority]:
+					// dispatched at least one task in this round
+					outstandingTasks = true
+
+					// submit will fail only when the processor has shut down,
+					// in which case we can just drop the task
+					_ = w.processor.Submit(task)
```

Comment:
do you need to handle the error?

Reply:
I think ignore the error and drop the task is the best option here.  The only error submit can and should return by processor is when it has been shutdown and failed to accept new tasks. In this case scheduler itself will also shutdown. We can Nack() the task and do a re-dispatch, but it will be a no-op as scheduler won't be able to accept it.

Actually I'd like to change the returned value for Submit from error to bool. Thoughts?

Reply:
Nack the task on error and added a unit test